File: tests/test_tcount_phasepoly.cc
// Tests for the experimental phase-polynomial T-count pass.

#include "clifft/backend/backend.h"
#include "clifft/circuit/parser.h"
#include "clifft/frontend/frontend.h"
#include "clifft/frontend/hir.h"
#include "clifft/optimizer/peephole.h"
#include "clifft/optimizer/tcount_phasepoly_pass.h"
#include "clifft/svm/svm.h"
#include "clifft/util/constants.h"

#include "test_helpers.h"

#include <algorithm>
#include <catch2/catch_test_macros.hpp>
#include <catch2/matchers/catch_matchers_floating_point.hpp>
#include <cmath>
#include <complex>
#include <string>
#include <vector>

using namespace clifft;
using clifft::test::X;
using clifft::test::Z;

static HirModule hir_from(const char* text) {
    return clifft::trace(clifft::parse(text));
}

// Dense statevector of a unitary circuit, optionally running the pass first.
static std::vector<std::complex<double>> statevector(const char* text, bool with_pass) {
    auto hir = hir_from(text);
    if (with_pass) {
        TCountPhasePolyPass pass;
        pass.run(hir);
    }
    auto prog = lower(hir);
    SchrodingerState state(prog.peak_rank, prog.total_meas_slots, prog.num_qubits);
    execute(prog, state);
    return get_statevector(prog, state);
}

// Assert the pass preserves the exact statevector (amplitudes AND global phase).
static void require_equiv(const char* text) {
    auto a = statevector(text, /*with_pass=*/false);
    auto b = statevector(text, /*with_pass=*/true);
    REQUIRE(a.size() == b.size());
    for (size_t i = 0; i < a.size(); ++i) {
        CHECK_THAT(b[i].real(), Catch::Matchers::WithinAbs(a[i].real(), 1e-9));
        CHECK_THAT(b[i].imag(), Catch::Matchers::WithinAbs(a[i].imag(), 1e-9));
    }
}

// =============================================================================
// Structural folding
// =============================================================================

TEST_CASE("PhasePoly: two T fold to a Clifford S (zero T)", "[tcount]") {
    auto hir = hir_from("T 0\nT 0");
    TCountPhasePolyPass pass;
    pass.run(hir);

    REQUIRE(hir.num_t_gates() == 0);
    REQUIRE(hir.ops.size() == 1);
    REQUIRE(hir.ops[0].op_type() == OpType::PHASE_ROTATION);
    REQUIRE(pass.blocks() == 1);
    REQUIRE(pass.t_before() == 2);
    REQUIRE(pass.t_after() == 0);
    REQUIRE(pass.t_removed() == 2);
}

TEST_CASE("PhasePoly: three T fold to one T plus Clifford", "[tcount]") {
    auto hir = hir_from("T 0\nT 0\nT 0");
    TCountPhasePolyPass pass;
    pass.run(hir);

    REQUIRE(hir.num_t_gates() == 1);
    REQUIRE(pass.t_before() == 3);
    REQUIRE(pass.t_after() == 1);
}

TEST_CASE("PhasePoly: four T fold to a Clifford Z (zero T)", "[tcount]") {
    auto hir = hir_from("T 0\nT 0\nT 0\nT 0");
    TCountPhasePolyPass pass;
    pass.run(hir);

    REQUIRE(hir.num_t_gates() == 0);
}

TEST_CASE("PhasePoly: eight T vanish to identity", "[tcount]") {
    auto hir = hir_from("T 0\nT 0\nT 0\nT 0\nT 0\nT 0\nT 0\nT 0");
    TCountPhasePolyPass pass;
    pass.run(hir);

    REQUIRE(hir.num_t_gates() == 0);
    REQUIRE(hir.ops.empty());
    REQUIRE(pass.t_removed() == 8);
}

TEST_CASE("PhasePoly: T and T_dag on same axis cancel", "[tcount]") {
    auto hir = hir_from("T 0\nT_DAG 0");
    TCountPhasePolyPass pass;
    pass.run(hir);

    REQUIRE(hir.ops.empty());
}

TEST_CASE("PhasePoly: two commuting axes fold independently", "[tcount]") {
    // Z(0) and Z(1) commute -> one block of 4, two axis groups, each c=2.
    auto hir = hir_from("T 0\nT 1\nT 0\nT 1");
    TCountPhasePolyPass pass;
    pass.run(hir);

    REQUIRE(hir.num_t_gates() == 0);
    REQUIRE(pass.blocks() == 1);
    REQUIRE(pass.max_block_axes() == 2);
    REQUIRE(pass.t_removed() == 4);
}

TEST_CASE("PhasePoly: anti-commuting axis blocks folding", "[tcount]") {
    // H rotates the middle T to the X axis, which anti-commutes with Z.
    // The three T gates split into singleton blocks and cannot fold.
    auto hir = hir_from("T 0\nH 0\nT 0\nH 0\nT 0");
    TCountPhasePolyPass pass;
    pass.run(hir);

    REQUIRE(hir.num_t_gates() == 3);
    REQUIRE(pass.blocks() == 0);
}

TEST_CASE("PhasePoly: single T is untouched", "[tcount]") {
    auto hir = hir_from("T 0");
    TCountPhasePolyPass pass;
    pass.run(hir);

    REQUIRE(hir.num_t_gates() == 1);
    REQUIRE(pass.blocks() == 0);
}

TEST_CASE("PhasePoly: multi-qubit ZZ axis folds", "[tcount]") {
    // CX entangles; T on qubit 1 then acts on a ZZ Pauli axis. Two such fold.
    auto hir = hir_from("CX 0 1\nT 1\nT 1");
    TCountPhasePolyPass pass;
    pass.run(hir);

    REQUIRE(hir.num_t_gates() == 0);
    REQUIRE(pass.blocks() == 1);
}

// =============================================================================
// Global-phase correctness (manual HIR, negative-sign axes)
// =============================================================================

TEST_CASE("PhasePoly: negative-sign T pair preserves global phase", "[tcount]") {
    // Two T on -Z(0): T(-Z) = exp(+i*pi/4) T_dag(+Z); the pair contributes a
    // net global phase of i and folds to S_dag.
    HirModule hir(1, /*pauli_capacity=*/16);
    clifft::test::append_tgate(hir, 0, Z(0), /*sign=*/true);
    clifft::test::append_tgate(hir, 0, Z(0), /*sign=*/true);
    auto initial_weight = hir.global_weight;

    TCountPhasePolyPass pass;
    pass.run(hir);

    REQUIRE(hir.num_t_gates() == 0);
    REQUIRE(hir.ops.size() == 1);
    REQUIRE(hir.ops[0].op_type() == OpType::PHASE_ROTATION);

    auto ratio = hir.global_weight / initial_weight;
    CHECK_THAT(ratio.real(), Catch::Matchers::WithinAbs(0.0, 1e-12));
    CHECK_THAT(ratio.imag(), Catch::Matchers::WithinAbs(1.0, 1e-12));
}

// =============================================================================
// Semantic equivalence via dense statevector (unitary circuits)
// =============================================================================

TEST_CASE("PhasePoly: statevector preserved on single-qubit T powers", "[tcount]") {
    require_equiv("H 0\nT 0\nT 0");
    require_equiv("H 0\nT 0\nT 0\nT 0");
    require_equiv("H 0\nT 0\nT 0\nT 0\nT 0\nT 0");
}

TEST_CASE("PhasePoly: statevector preserved on entangled axes", "[tcount]") {
    require_equiv("H 0\nH 1\nCX 0 1\nT 1\nT 1\nT 1");
    require_equiv("H 0\nT 0\nT 1\nT 0\nT 1\nCX 0 1");
}

TEST_CASE("PhasePoly: statevector preserved with interleaved different axes", "[tcount]") {
    require_equiv("H 0\nH 1\nH 2\nT 0\nT 1\nT 2\nT 0\nT 1\nT 2");
}

// =============================================================================
// Phase B (TOHPE) end-to-end: a real multi-axis reduction with the diagonal
// unitary verified exactly. This is the evidence the issue asks for -- TOHPE
// removing T gates that phase folding / peephole cannot, semantics preserved.
// =============================================================================

// "T on Z-parity a" (a's set bits), computed onto the lowest support qubit by a
// CNOT ladder so the front end records a single virtual Z-parity T.
static std::string t_on_parity(uint32_t a, int nq) {
    std::vector<int> bits;
    for (int q = 0; q < nq; ++q)
        if (a & (1u << q))
            bits.push_back(q);
    int tgt = bits[0];
    std::string s;
    for (size_t k = 1; k < bits.size(); ++k)
        s += "CX " + std::to_string(bits[k]) + " " + std::to_string(tgt) + "\n";
    s += "T " + std::to_string(tgt) + "\n";
    for (size_t k = bits.size(); k-- > 1;)
        s += "CX " + std::to_string(bits[k]) + " " + std::to_string(tgt) + "\n";
    return s;
}

// Diagonal phase function f(x) mod 8 of a purely Z-type HIR (T_GATE and
// PHASE_ROTATION ops only), evaluated over all 2^nq inputs. This is the exact
// unitary the block implements, independent of how it is factored.
static std::vector<int> hir_phase_fn(const HirModule& hir, int nq) {
    std::vector<int> f(1u << nq, 0);
    for (const auto& op : hir.ops) {
        int coeff;
        if (op.op_type() == OpType::T_GATE) {
            coeff = op.is_dagger() ? -1 : 1;
        } else if (op.op_type() == OpType::PHASE_ROTATION) {
            coeff = static_cast<int>(std::llround(op.alpha() * 4.0));
        } else {
            continue;
        }
        REQUIRE(hir.destab_mask(op).is_zero());  // Z-only
        REQUIRE(!hir.sign(op));
        uint64_t parity = hir.stab_mask(op).words[0];
        for (int x = 0; x < (1 << nq); ++x) {
            int bit = static_cast<int>(__builtin_popcountll(parity & static_cast<uint64_t>(x)) & 1);
            f[x] = (((f[x] + coeff * bit) % 8) + 8) % 8;
        }
    }
    return f;
}

TEST_CASE("PhasePoly TOHPE: S_empty (15 parities) reduces, diagonal preserved", "[tcount][tohpe]") {
    // Amy-Maslov-Mosca arXiv:1303.2042: T on every nonzero parity of F_2^4 is
    // the identity (f = 0). Phase folding removes none (15 distinct axes); TOHPE
    // removes T gates while preserving the exact diagonal unitary.
    std::string text;
    for (uint32_t a = 1; a < 16; ++a)
        text += t_on_parity(a, 4);

    auto hir = hir_from(text.c_str());
    REQUIRE(hir.num_t_gates() == 15);
    auto f_before = hir_phase_fn(hir, 4);

    TCountPhasePolyPass pass;
    pass.run(hir);

    INFO("t_after=" << hir.num_t_gates() << " tohpe_removed=" << pass.tohpe_removed());
    REQUIRE(hir.num_t_gates() < 15);  // genuine multi-axis reduction
    REQUIRE(pass.tohpe_removed() > 0);
    REQUIRE(pass.tohpe_blocks() >= 1);
    REQUIRE(hir_phase_fn(hir, 4) == f_before);  // exact semantics
}

TEST_CASE("PhasePoly TOHPE: 14-parity block collapses to one T, diagonal preserved",
          "[tcount][tohpe]") {
    // The 15 parities minus Z_0123 implement a single T_dag on Z_0^Z_1^Z_2^Z_3:
    // a nontrivial diagonal unitary that TOHPE should reduce from 14 T to ~1 T.
    std::string text;
    for (uint32_t a = 1; a < 15; ++a)  // skip a = 15 (Z_0123)
        text += t_on_parity(a, 4);

    auto hir = hir_from(text.c_str());
    REQUIRE(hir.num_t_gates() == 14);
    auto f_before = hir_phase_fn(hir, 4);
    // Confirm the target is the nontrivial single-T_dag phase.
    for (int x = 0; x < 16; ++x) {
        int p = __builtin_popcount(x & 0xF) & 1;
        REQUIRE(f_before[x] == (p ? 7 : 0));
    }

    TCountPhasePolyPass pass;
    pass.run(hir);

    INFO("t_after=" << hir.num_t_gates() << " tohpe_removed=" << pass.tohpe_removed());
    REQUIRE(hir.num_t_gates() < 14);
    REQUIRE(hir.num_t_gates() >= 1);
    REQUIRE(pass.tohpe_removed() > 0);
    REQUIRE(hir_phase_fn(hir, 4) == f_before);  // exact semantics
}

// CCZ(a,b,c) as its 7-term phase polynomial (+singles, -pairs, +triple).
static std::string ccz_str(int a, int b, int c) {
    auto par = [](uint32_t mask, bool dag) {
        std::vector<int> bits;
        for (int q = 0; q < 12; ++q)
            if (mask & (1u << q))
                bits.push_back(q);
        int t = bits[0];
        std::string s;
        for (size_t k = 1; k < bits.size(); ++k)
            s += "CX " + std::to_string(bits[k]) + " " + std::to_string(t) + "\n";
        s += (dag ? "T_DAG " : "T ") + std::to_string(t) + "\n";
        for (size_t k = bits.size(); k-- > 1;)
            s += "CX " + std::to_string(bits[k]) + " " + std::to_string(t) + "\n";
        return s;
    };
    uint32_t A = 1u << a, B = 1u << b, C = 1u << c;
    return par(A, false) + par(B, false) + par(C, false) + par(A | B, true) + par(A | C, true) +
           par(B | C, true) + par(A | B | C, false);
}

TEST_CASE("PhasePoly TOHPE: dense CCZ-complete block reduces beyond peephole, exact",
          "[tcount][tohpe]") {
    // All C(6,3)=20 CCZ triples on 6 qubits: a dense diagonal phase polynomial.
    // Peephole/folding reach 20 T; faithful TOHPE removes more (a real
    // ancilla-free multi-axis reduction on structured input), exactly.
    std::string text;
    for (int a = 0; a < 6; ++a)
        for (int b = a + 1; b < 6; ++b)
            for (int c = b + 1; c < 6; ++c)
                text += ccz_str(a, b, c);

    auto hir_peep = hir_from(text.c_str());
    PeepholeFusionPass().run(hir_peep);
    size_t t_peephole = hir_peep.num_t_gates();
    auto f_before = hir_phase_fn(hir_peep, 6);

    auto hir = hir_from(text.c_str());
    PeepholeFusionPass().run(hir);
    TCountPhasePolyPass pass;
    pass.run(hir);

    INFO("t_peephole=" << t_peephole << " t_after=" << hir.num_t_gates()
                       << " tohpe_removed=" << pass.tohpe_removed());
    REQUIRE(hir.num_t_gates() < t_peephole);  // TOHPE beats peephole here
    REQUIRE(pass.tohpe_removed() > 0);
    REQUIRE(hir_phase_fn(hir, 6) == f_before);  // exact diagonal unitary
}

TEST_CASE("PhasePoly TOHPE: mixed-type blocks are left untouched and exact", "[tcount][tohpe]") {
    // Phase B reduces only single-Pauli-type (diagonal) blocks, where f(x) mod 8
    // is the exact unitary. A block whose axes mix X and Z -- here s_empty
    // conjugated by a Clifford -- must be left for plain folding; reducing it
    // would need quadratic-Clifford / global-phase tracking the f-check does not
    // provide. This pins the regression a bell-mixed conjugation exposed.
    std::string inner;
    for (uint32_t a = 1; a < 16; ++a)
        inner += t_on_parity(a, 4);  // s_empty(4): trivial diagonal phase

    const std::string h_only = "H 0\n" + inner + "H 0\n";
    const std::string bell = "CX 0 1\nH 0\n" + inner + "H 0\nCX 0 1\n";

    for (const auto& text : {h_only, bell}) {
        auto hir = hir_from(text.c_str());
        TCountPhasePolyPass pass;
        pass.run(hir);
        REQUIRE(pass.tohpe_removed() == 0);  // mixed block skipped, not reduced
        require_equiv(text.c_str());         // unitary preserved exactly
    }
}

// Every source line referenced by the current mapping, as a flat list.
static std::vector<uint32_t> all_source_lines(const HirModule& hir) {
    std::vector<uint32_t> v;
    for (const auto& lines : hir.source_map)
        for (uint32_t ln : lines)
            v.push_back(ln);
    return v;
}

TEST_CASE("PhasePoly: source map stays consistent after reduction", "[tcount][sourcemap]") {
    // T on 14 of the 15 nonzero parities of F_2^4: TOHPE collapses it to a
    // single T, deleting most ops and re-emitting the survivor as a different
    // Pauli. The source map must stay parallel to ops, reference only lines the
    // input already had, and consolidate the deleted ops' lines onto a survivor.
    std::string text;
    for (uint32_t a = 1; a < 15; ++a)  // skip a == 15 (full parity)
        text += t_on_parity(a, 4);

    auto hir = hir_from(text.c_str());
    REQUIRE(hir.source_map.size() == hir.ops.size());
    const std::vector<uint32_t> orig = all_source_lines(hir);
    const size_t t_before = hir.num_t_gates();

    TCountPhasePolyPass().run(hir);

    REQUIRE(hir.num_t_gates() < t_before);             // the block actually reduced
    REQUIRE(hir.source_map.size() == hir.ops.size());  // parallel after compaction
    size_t max_entries = 0;
    for (const auto& lines : hir.source_map) {
        for (uint32_t ln : lines)
            REQUIRE(std::find(orig.begin(), orig.end(), ln) != orig.end());  // no new line
        max_entries = std::max(max_entries, lines.size());
    }
    // The block's deleted ops are not dropped: their lines are consolidated onto
    // a surviving op, so at least one op carries more than its own single line.
    // (Without the reattach every op would map to exactly one line.)
    REQUIRE(max_entries >= 2);
}

TEST_CASE("PhasePoly: source map consistent after pure folding", "[tcount][sourcemap]") {
    // Three T on one axis fold to one T plus an S. No TOHPE fires, so this
    // exercises the folding re-emit path's source-map handling.
    auto hir = hir_from("T 0\nT 0\nT 0");
    REQUIRE(hir.source_map.size() == hir.ops.size());
    const std::vector<uint32_t> orig = all_source_lines(hir);

    TCountPhasePolyPass().run(hir);

    REQUIRE(hir.source_map.size() == hir.ops.size());  // parallel after compaction
    for (const auto& lines : hir.source_map)
        for (uint32_t ln : lines)
            REQUIRE(std::find(orig.begin(), orig.end(), ln) != orig.end());  // no new line
}

TEST_CASE("PhasePoly: max_verify_bits gates Phase B", "[tcount][tohpe]") {
    // S_empty(4) spans support 4. A verify cutoff below 4 leaves Phase B unable
    // to check the move, so the block is returned unchanged; the default reduces
    // it. This pins down that the parameter actually controls the behaviour.
    std::string text;
    for (uint32_t a = 1; a < 16; ++a)
        text += t_on_parity(a, 4);

    auto low = hir_from(text.c_str());
    TCountPhasePolyPass(/*enable_tohpe=*/true, /*max_verify_bits=*/3).run(low);
    REQUIRE(low.num_t_gates() == 15);  // cutoff below the support: no reduction

    auto def = hir_from(text.c_str());
    TCountPhasePolyPass(/*enable_tohpe=*/true, /*max_verify_bits=*/14).run(def);
    REQUIRE(def.num_t_gates() < 15);  // default cutoff: collapses the block
}

File: tests/python/test_tcount_phasepoly.py
"""Correctness + behaviour tests for the experimental TCountPhasePolyPass.

Two verification strategies, mirroring test_peephole_oracle.py:

1. Self-consistency: compile with the pass on vs off, assert identical
   statevectors (fidelity ~ 1), proving exact semantics preservation.
2. Structural: the pass folds per-axis T coefficients in Z_8 and reports
   analyzer metrics; we check the metrics and the known algebraic identities.

Phase A folding alone reaches the same per-block optimum PeepholeFusionPass
already reaches, so after peephole it removes nothing further. Phase B (TOHPE)
goes below that on single-type diagonal blocks with cubic redundancy: see
TestTohpe, where s_empty collapses to 0 T.
"""

import numpy as np
import pytest
from conftest import (
    assert_statevectors_equal,
    random_clifford_t_circuit,
    random_dense_clifford_t_circuit,
)

import clifft


def _phasepoly_only(circuit_str: str) -> clifft.Program:
    """Compile running ONLY the experimental pass on the HIR."""
    hir = clifft.trace(clifft.parse(circuit_str))
    pm = clifft.HirPassManager()
    pm.add(clifft.TCountPhasePolyPass())
    pm.run(hir)
    return clifft.lower(hir)


def _statevector(prog: clifft.Program) -> np.ndarray:
    state = clifft.State(peak_rank=prog.peak_rank, num_measurements=prog.num_measurements)
    clifft.execute(prog, state)
    return np.array(clifft.get_statevector(prog, state))


def _t_count(circuit_str: str, *, optimize: bool) -> int:
    hir = clifft.trace(clifft.parse(circuit_str))
    if optimize:
        pm = clifft.HirPassManager()
        pm.add(clifft.TCountPhasePolyPass())
        pm.run(hir)
    return int(hir.num_t_gates)


# ---------------------------------------------------------------------------
# Semantic equivalence (pass on vs off) on random Clifford+T circuits
# ---------------------------------------------------------------------------


class TestStatevectorEquivalence:
    @pytest.mark.parametrize("num_qubits", [2, 3, 4, 5, 6])
    @pytest.mark.parametrize("seed", range(5))
    def test_random_small(self, num_qubits: int, seed: int) -> None:
        circuit = random_clifford_t_circuit(num_qubits, depth=20, seed=seed)
        base = _statevector(clifft.compile(circuit, hir_passes=None, bytecode_passes=None))
        opt = _statevector(_phasepoly_only(circuit))
        assert_statevectors_equal(opt, base, msg=f"{num_qubits}q seed={seed}")

    @pytest.mark.parametrize("seed", range(8))
    def test_random_8q_depth30(self, seed: int) -> None:
        circuit = random_clifford_t_circuit(8, depth=30, seed=seed)
        base = _statevector(clifft.compile(circuit, hir_passes=None, bytecode_passes=None))
        opt = _statevector(_phasepoly_only(circuit))
        assert_statevectors_equal(opt, base, msg=f"8q seed={seed}")

    @pytest.mark.parametrize("seed", range(5))
    def test_dense_entangled(self, seed: int) -> None:
        circuit = random_dense_clifford_t_circuit(5, depth=40, seed=seed)
        base = _statevector(clifft.compile(circuit, hir_passes=None, bytecode_passes=None))
        opt = _statevector(_phasepoly_only(circuit))
        assert_statevectors_equal(opt, base, msg=f"dense 5q seed={seed}")


# ---------------------------------------------------------------------------
# Algebraic identities and analyzer metrics
# ---------------------------------------------------------------------------


class TestFolding:
    def test_two_t_to_clifford(self) -> None:
        assert _t_count("T 0\nT 0", optimize=False) == 2
        assert _t_count("T 0\nT 0", optimize=True) == 0

    def test_three_t_to_one(self) -> None:
        assert _t_count("T 0\nT 0\nT 0", optimize=True) == 1

    def test_four_t_to_clifford(self) -> None:
        assert _t_count("T 0\nT 0\nT 0\nT 0", optimize=True) == 0

    def test_eight_t_to_identity(self) -> None:
        assert _t_count("T 0\nT 0\nT 0\nT 0\nT 0\nT 0\nT 0\nT 0", optimize=True) == 0

    def test_anticommuting_axis_blocks(self) -> None:
        # H rotates the middle T off the Z axis; the three T gates cannot fold.
        assert _t_count("T 0\nH 0\nT 0\nH 0\nT 0", optimize=True) == 3

    def test_pass_metrics(self) -> None:
        hir = clifft.trace(clifft.parse("T 0\nT 0\nT 1\nT 1\nT 1"))
        p = clifft.TCountPhasePolyPass()
        pm = clifft.HirPassManager()
        pm.add(p)
        pm.run(hir)
        # One commuting block of 5 (Z0,Z1 commute); two axis groups.
        assert p.blocks == 1
        assert p.t_before == 5
        assert p.max_block_axes == 2
        # q0: c=2 -> 0 T; q1: c=3 -> 1 T. Net 4 removed.
        assert p.t_after == 1
        assert p.t_removed == 4


class TestTohpe:
    """Phase B (multi-axis TOHPE) removes T gates that folding cannot."""

    @staticmethod
    def _parity(mask: int, dag: bool = False) -> str:
        bits = [q for q in range(8) if mask & (1 << q)]
        t = bits[0]
        up = "".join(f"CX {b} {t}\n" for b in bits[1:])
        dn = "".join(f"CX {b} {t}\n" for b in reversed(bits[1:]))
        return up + ("T_DAG " if dag else "T ") + f"{t}\n" + dn

    def test_s_empty_reduces_beyond_folding(self) -> None:
        # T on every nonzero parity of F_2^4 (Amy-Maslov-Mosca S_empty): folding
        # removes none (15 distinct axes); TOHPE collapses it.
        text = "".join(self._parity(a) for a in range(1, 16))
        hir = clifft.trace(clifft.parse(text))
        assert int(hir.num_t_gates) == 15

        p = clifft.TCountPhasePolyPass()  # TOHPE enabled
        pm = clifft.HirPassManager()
        pm.add(p)
        pm.run(hir)

        assert int(hir.num_t_gates) < 15
        assert p.tohpe_removed > 0
        assert p.tohpe_blocks >= 1

    def test_folding_alone_does_not_reduce_s_empty(self) -> None:
        text = "".join(self._parity(a) for a in range(1, 16))
        hir = clifft.trace(clifft.parse(text))
        p = clifft.TCountPhasePolyPass(enable_tohpe=False)
        pm = clifft.HirPassManager()
        pm.add(p)
        pm.run(hir)
        assert int(hir.num_t_gates) == 15  # folding sees 15 distinct axes

    def test_mixed_type_block_left_untouched_and_exact(self) -> None:
        # Phase B reduces only single-type diagonal blocks. s_empty conjugated by
        # an entangling Clifford is mixed-type and must be left for plain folding;
        # reducing it would need quadratic-Clifford / global-phase tracking the
        # f-check does not provide. Compared componentwise (not by fidelity) so a
        # dropped global phase would fail.
        inner = "".join(self._parity(a) for a in range(1, 16))  # s_empty(4)
        text = "CX 0 1\nH 0\n" + inner + "H 0\nCX 0 1\n"

        hir = clifft.trace(clifft.parse(text))
        p = clifft.TCountPhasePolyPass()
        pm = clifft.HirPassManager()
        pm.add(p)
        pm.run(hir)
        assert p.tohpe_removed == 0  # mixed block skipped, not reduced

        opt = _statevector(_phasepoly_only(text))
        base = _statevector(clifft.compile(text, hir_passes=None, bytecode_passes=None))
        assert float(np.max(np.abs(opt - base))) < 1e-9  # exact, incl. global phase

    def test_max_verify_bits_kwarg_gates_reduction(self) -> None:
        # S_empty(4) spans support 4. A cutoff below 4 disables Phase B; the
        # default reduces. Checks the binding's max_verify_bits argument.
        text = "".join(self._parity(a) for a in range(1, 16))

        def t_count(max_verify_bits: int) -> int:
            hir = clifft.trace(clifft.parse(text))
            pm = clifft.HirPassManager()
            pm.add(clifft.TCountPhasePolyPass(max_verify_bits=max_verify_bits))
            pm.run(hir)
            return int(hir.num_t_gates)

        assert t_count(3) == 15  # cutoff below the support: unchanged
        assert t_count(14) < 15  # default cutoff: reduces


# ---------------------------------------------------------------------------
# Central finding: peephole already reaches the per-block optimum
# ---------------------------------------------------------------------------


class TestRedundantWithPeephole:
    @pytest.mark.parametrize("seed", range(8))
    def test_no_gain_after_peephole(self, seed: int) -> None:
        """After PeepholeFusionPass, the pass finds nothing further to remove.

        Demonstrates that same-axis folding (which peephole already performs)
        is the commuting-block optimum: any leftover multi-axis reduction needs
        structure outside the current HIR.
        """
        circuit = random_clifford_t_circuit(6, depth=30, seed=seed)
        hir = clifft.trace(clifft.parse(circuit))
        pm = clifft.HirPassManager()
        pm.add(clifft.PeepholeFusionPass())
        pm.run(hir)
        t_after_peephole = int(hir.num_t_gates)

        # Phase A folding only (enable_tohpe=False): this isolates the claim
        # that same-axis folding adds nothing once peephole has run.
        pp = clifft.TCountPhasePolyPass(enable_tohpe=False)
        pm2 = clifft.HirPassManager()
        pm2.add(pp)
        pm2.run(hir)

        assert pp.t_removed == 0
        assert int(hir.num_t_gates) == t_after_peephole


# ---------------------------------------------------------------------------
# Pass enabled in the default pipeline: a "default-on" run over the corpus,
# confirming it composes with the other passes and preserves the circuit.
# ---------------------------------------------------------------------------


class TestPassInDefaultPipeline:
    @staticmethod
    def _with_pass(circuit_str: str) -> clifft.Program:
        """Default HIR pipeline (peephole, squeeze) with the pass inserted."""
        hir = clifft.trace(clifft.parse(circuit_str))
        pm = clifft.HirPassManager()
        pm.add(clifft.PeepholeFusionPass())
        pm.add(clifft.TCountPhasePolyPass())
        pm.add(clifft.StatevectorSqueezePass())
        pm.run(hir)
        return clifft.lower(hir)

    @pytest.mark.parametrize("num_qubits", [2, 3, 4, 5, 6])
    @pytest.mark.parametrize("seed", range(6))
    def test_random_matches_default(self, num_qubits: int, seed: int) -> None:
        circuit = random_clifford_t_circuit(num_qubits, depth=30, seed=seed)
        base = _statevector(clifft.compile(circuit))  # default passes
        opt = _statevector(self._with_pass(circuit))
        assert_statevectors_equal(opt, base, msg=f"{num_qubits}q seed={seed}")

    @pytest.mark.parametrize("seed", range(5))
    def test_dense_entangled_matches_default(self, seed: int) -> None:
        circuit = random_dense_clifford_t_circuit(5, depth=40, seed=seed)
        base = _statevector(clifft.compile(circuit))
        opt = _statevector(self._with_pass(circuit))
        assert_statevectors_equal(opt, base, msg=f"dense 5q seed={seed}")
